// loas3/dist/validator/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ajv_1 = __importDefault(require("ajv"));
var ajv_errors_1 = __importDefault(require("ajv-errors"));
var ajv_keywords_1 = __importDefault(require("ajv-keywords"));
var full_1 = __importDefault(require("../schema/full"));
var lazy_1 = __importDefault(require("../schema/lazy"));
var Validator = /** @class */ (function () {
    function Validator() {
        this.ajv = new ajv_1.default({
            allErrors: true,
            jsonPointers: true
        });
        ajv_keywords_1.default(this.ajv, "switch");
        ajv_errors_1.default(this.ajv);
        this.addSchema(full_1.default, "oas3");
        this.addSchema(lazy_1.default, "loas3");
    }
    Validator.prototype.addSchema = function (schema, key) {
        this.ajv.addSchema(schema, key);
    };
    Validator.prototype.validate = function (key, data) {
        this.ajv.validate(key, data);
        return this.ajv.errors;
    };
    return Validator;
}());
exports.default = Validator;
