// io-ts/es6/Encoder.js
import { intersect } from './Decoder';
import { memoize } from './Schemable';
import { identity } from 'fp-ts/es6/function';
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export function id() {
    return {
        encode: identity
    };
}
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export function nullable(or) {
    return {
        encode: function (a) { return (a === null ? null : or.encode(a)); }
    };
}
/**
 * @since 2.2.3
 */
export function type(properties) {
    return {
        encode: function (a) {
            var o = {};
            for (var k in properties) {
                o[k] = properties[k].encode(a[k]);
            }
            return o;
        }
    };
}
/**
 * @since 2.2.3
 */
export function partial(properties) {
    return {
        encode: function (a) {
            var o = {};
            for (var k in properties) {
                var v = a[k];
                // don't add missing properties
                if (k in a) {
                    // don't strip undefined properties
                    o[k] = v === undefined ? undefined : properties[k].encode(v);
                }
            }
            return o;
        }
    };
}
/**
 * @since 2.2.3
 */
export function record(codomain) {
    return {
        encode: function (r) {
            var o = {};
            for (var k in r) {
                o[k] = codomain.encode(r[k]);
            }
            return o;
        }
    };
}
/**
 * @since 2.2.3
 */
export function array(items) {
    return {
        encode: function (as) { return as.map(items.encode); }
    };
}
/**
 * @since 2.2.3
 */
export function tuple() {
    var components = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        components[_i] = arguments[_i];
    }
    return {
        encode: function (as) { return components.map(function (c, i) { return c.encode(as[i]); }); }
    };
}
/**
 * @since 2.2.3
 */
export function intersection(left, right) {
    return {
        encode: function (ab) { return intersect(left.encode(ab), right.encode(ab)); }
    };
}
/**
 * @since 2.2.3
 */
export function sum(tag) {
    return function (members) {
        return {
            encode: function (a) { return members[a[tag]].encode(a); }
        };
    };
}
/**
 * @since 2.2.3
 */
export function lazy(f) {
    var get = memoize(f);
    return {
        encode: function (a) { return get().encode(a); }
    };
}
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var contramap = function (f) { return function (fa) {
    return contramap_(fa, f);
}; };
var contramap_ = function (fa, f) { return ({
    encode: function (b) { return fa.encode(f(b)); }
}); };
/**
 * @since 2.2.3
 */
export var compose = function (ea) { return function (ab) {
    return compose_(ab, ea);
}; };
var compose_ = function (ab, ea) { return ({
    encode: function (b) { return ea.encode(ab.encode(b)); }
}); };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var URI = 'io-ts/Encoder';
/**
 * @since 2.2.3
 */
export var contravariantEncoder = {
    URI: URI,
    contramap: contramap_
};
/**
 * @since 2.2.3
 */
export var categoryEncoder = {
    URI: URI,
    compose: compose_,
    id: id
};
