// io-ts/es6/Eq.js
/**
 * @since 2.2.2
 */
import * as A from 'fp-ts/es6/Array';
import * as E from 'fp-ts/es6/Eq';
import * as R from 'fp-ts/es6/Record';
import { memoize } from './Schemable';
// -------------------------------------------------------------------------------------
// primitives
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.2
 */
export var string = E.eqString;
/**
 * @since 2.2.2
 */
export var number = E.eqNumber;
/**
 * @since 2.2.2
 */
export var boolean = E.eqBoolean;
/**
 * @since 2.2.2
 */
export var UnknownArray = E.fromEquals(function (x, y) { return x.length === y.length; });
/**
 * @since 2.2.2
 */
export var UnknownRecord = E.fromEquals(function (x, y) {
    for (var k in x) {
        if (!(k in y)) {
            return false;
        }
    }
    for (var k in y) {
        if (!(k in x)) {
            return false;
        }
    }
    return true;
});
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.2
 */
export function nullable(or) {
    return {
        equals: function (x, y) { return (x === null || y === null ? x === y : or.equals(x, y)); }
    };
}
/**
 * @since 2.2.2
 */
export var type = E.getStructEq;
/**
 * @since 2.2.2
 */
export function partial(properties) {
    return {
        equals: function (x, y) {
            for (var k in properties) {
                var xk = x[k];
                var yk = y[k];
                if (!(xk === undefined || yk === undefined ? xk === yk : properties[k].equals(xk, yk))) {
                    return false;
                }
            }
            return true;
        }
    };
}
/**
 * @since 2.2.2
 */
export var record = R.getEq;
/**
 * @since 2.2.2
 */
export var array = A.getEq;
/**
 * @since 2.2.2
 */
export var tuple = E.getTupleEq;
/**
 * @since 2.2.2
 */
export function intersection(left, right) {
    return {
        equals: function (x, y) { return left.equals(x, y) && right.equals(x, y); }
    };
}
/**
 * @since 2.2.2
 */
export function sum(tag) {
    return function (members) {
        return {
            equals: function (x, y) {
                var vx = x[tag];
                var vy = y[tag];
                if (vx !== vy) {
                    return false;
                }
                return members[vx].equals(x, y);
            }
        };
    };
}
/**
 * @since 2.2.2
 */
export function lazy(f) {
    var get = memoize(f);
    return {
        equals: function (x, y) { return get().equals(x, y); }
    };
}
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var schemableEq = {
    URI: E.URI,
    literal: function () { return E.eqStrict; },
    string: string,
    number: number,
    boolean: boolean,
    nullable: nullable,
    type: type,
    partial: partial,
    record: record,
    array: array,
    tuple: tuple,
    intersection: intersection,
    sum: sum,
    lazy: function (_, f) { return lazy(f); },
    UnknownArray: UnknownArray,
    UnknownRecord: UnknownRecord,
    refinement: function (from) { return from; }
};
