// io-ts/es6/JsonEncoder.js
import { identity } from 'fp-ts/es6/function';
import * as E from './Encoder';
// -------------------------------------------------------------------------------------
// primitives
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var id = {
    encode: identity
};
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var nullable = E.nullable;
/**
 * @since 2.2.3
 */
export var type = E.type;
/**
 * @since 2.2.3
 */
export var partial = E.partial;
/**
 * @since 2.2.3
 */
export var record = E.record;
/**
 * @since 2.2.3
 */
export var array = E.array;
/**
 * @since 2.2.3
 */
export var tuple = E.tuple;
/**
 * @since 2.2.3
 */
export var intersection = E.intersection;
/**
 * @since 2.2.3
 */
export var sum = E.sum;
/**
 * @since 2.2.3
 */
export var lazy = E.lazy;
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var contramap = E.contramap;
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var URI = 'io-ts/JsonEncoder';
/**
 * @since 2.2.3
 */
export var contravariantJsonEncoder = {
    URI: URI,
    contramap: E.contravariantEncoder.contramap
};
/**
 * @since 2.2.3
 */
export var schemableJsonEncoder = {
    URI: URI,
    literal: function () { return id; },
    string: id,
    number: id,
    boolean: id,
    nullable: nullable,
    type: type,
    partial: partial,
    record: record,
    array: array,
    tuple: tuple,
    intersection: intersection,
    sum: sum,
    lazy: function (_, f) { return lazy(f); }
};
