// io-ts/es6/JsonCodec.js
import * as C from './Codec';
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var make = C.make;
/**
 * @since 2.2.3
 */
export var literal = C.literal;
// -------------------------------------------------------------------------------------
// primitives
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var string = C.string;
/**
 * @since 2.2.3
 */
export var number = C.number;
/**
 * @since 2.2.3
 */
export var boolean = C.boolean;
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var withExpected = C.withExpected;
/**
 * @since 2.2.3
 */
export var refinement = C.refinement;
/**
 * @since 2.2.3
 */
export var nullable = C.nullable;
/**
 * @since 2.2.3
 */
export var type = C.type;
/**
 * @since 2.2.3
 */
export var partial = C.partial;
/**
 * @since 2.2.3
 */
export var record = C.record;
/**
 * @since 2.2.3
 */
export var array = C.array;
/**
 * @since 2.2.3
 */
export var tuple = C.tuple;
/**
 * @since 2.2.3
 */
export var intersection = C.intersection;
/**
 * @since 2.2.3
 */
export var sum = C.sum;
/**
 * @since 2.2.3
 */
export var lazy = C.lazy;
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var imap = C.imap;
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
export var URI = 'io-ts/JsonCodec';
/**
 * @since 2.2.3
 */
export var invariantJsonCodec = {
    URI: URI,
    imap: C.invariantCodec.imap
};
/**
 * @since 2.2.3
 */
export var schemableJsonCodec = {
    URI: URI,
    literal: literal,
    string: string,
    number: number,
    boolean: boolean,
    nullable: nullable,
    type: type,
    partial: partial,
    record: record,
    array: array,
    tuple: tuple,
    intersection: intersection,
    sum: sum,
    lazy: lazy,
    refinement: refinement
};
