// json-schema-poet/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nul_ = function (u) { return function () { return (__assign({ type: "null" }, u)); }; };
exports.nul = exports.nul_({});
exports.cnst_ = function (u) { return function (c) { return (__assign({ const: c }, u)); }; };
exports.cnst = exports.cnst_({});
exports.integer_ = function (u) { return function (props) { return (__assign(__assign({ type: "integer" }, (props || {})), u)); }; };
exports.integer = exports.integer_({});
exports.number_ = function (u) { return function (props) { return (__assign(__assign({ type: "number" }, (props || {})), u)); }; };
exports.number = exports.number_({});
exports.string_ = function (u) { return function (faker) { return (__assign(__assign({ type: "string" }, (faker !== undefined ? { faker: faker } : {})), u)); }; };
exports.string = exports.string_({});
exports.stringEnum_ = function (u) { return function (enu) { return (__assign({ type: "string", enum: enu }, u)); }; };
exports.stringEnum = exports.stringEnum_({});
exports.numberEnum_ = function (u) { return function (enu) { return (__assign({ type: "number", enum: enu }, u)); }; };
exports.numberEnum = exports.numberEnum_({});
exports.integerEnum_ = function (u) { return function (enu) { return (__assign({ type: "integer", enum: enu }, u)); }; };
exports.integerEnum = exports.integerEnum_({});
exports.$ref_ = function (u) { return function ($ref) { return (__assign({ $ref: $ref }, u)); }; };
exports.$ref = exports.$ref_({});
exports.regex_ = function (u) { return function (pattern) { return (__assign({ type: "string", pattern: pattern }, u)); }; };
exports.regex = exports.regex_({});
exports.boolean_ = function (u) { return function () { return (__assign({ type: "boolean" }, u)); }; };
exports.boolean = exports.boolean_({});
exports.array_ = function (u) { return function (items) { return (__assign({ type: "array", items: items }, u)); }; };
exports.array = function (items) { return exports.array_({})(items); };
exports.tuple_ = function (u) { return function (items) { return (__assign({ type: "array", items: items }, u)); }; };
exports.tuple = function (items) { return exports.tuple_({})(items); };
exports.allOf_ = function (u) { return function (allOf) { return (__assign({ allOf: allOf }, u)); }; };
exports.allOf = function (allOf) { return exports.allOf_({})(allOf); };
exports.anyOf_ = function (u) { return function (anyOf) { return (__assign({ anyOf: anyOf }, u)); }; };
exports.anyOf = function (anyOf) { return exports.anyOf_({})(anyOf); };
exports.oneOf_ = function (u) { return function (oneOf) { return (__assign({ oneOf: oneOf }, u)); }; };
exports.oneOf = function (oneOf) { return exports.oneOf_({})(oneOf); };
exports.not_ = function (u) { return function (not) { return (__assign({ not: not }, u)); }; };
exports.not = function (not) { return exports.not_({})(not); };
exports.dictionary_ = function (u) { return function (vals) { return exports.object_(u)({ additionalProperties: vals }); }; };
exports.dictionary = function (vals) {
    return exports.dictionary_({})(vals);
};
exports.type_ = function (u) { return function (req, opt) {
    return exports.object_(u)({
        properties: __assign(__assign({}, req), opt),
        required: Object.keys(req)
    });
}; };
exports.type = function (req, opt) { return exports.type_({})(req, opt); };
exports.object_ = function (u) { return function (props) { return (__assign(__assign(__assign(__assign(__assign({ type: "object" }, (props && props.required ? { required: props.required } : {})), (props && props.properties
    ? {
        properties: Object.entries(props.properties)
            .map(function (_a) {
            var _b;
            var _c = __read(_a, 2), a = _c[0], b = _c[1];
            return (_b = {}, _b[a] = b, _b);
        })
            .reduce(function (a, b) { return (__assign(__assign({}, a), b)); }, {})
    }
    : {})), (props && props.patternProperties
    ? {
        patternProperties: Object.entries(props.patternProperties)
            .map(function (_a) {
            var _b;
            var _c = __read(_a, 2), a = _c[0], b = _c[1];
            return (_b = {}, _b[a] = b, _b);
        })
            .reduce(function (a, b) { return (__assign(__assign({}, a), b)); }, {})
    }
    : {})), (props && props.additionalProperties
    ? {
        additionalProperties: typeof props.additionalProperties === "boolean"
            ? props.additionalProperties
            : props.additionalProperties
    }
    : {})), u)); }; };
exports.object = function (props) {
    return exports.object_({})(props);
};
exports.top = function (anything, props) { return (__assign(__assign({}, anything), props)); };
exports.extend = function (what, key, v) {
    var _a;
    return (__assign(__assign({}, what), (_a = {}, _a[key] = v, _a)));
};
exports.extendT = function (u) { return ({
    nul: exports.nul_(u),
    cnst: exports.cnst_(u),
    integer: exports.integer_(u),
    number: exports.number_(u),
    string: exports.string_(u),
    stringEnum: exports.stringEnum_(u),
    numberEnum: exports.numberEnum_(u),
    integerEnum: exports.integerEnum_(u),
    regex: exports.regex_(u),
    boolean: exports.boolean_(u),
    $ref: exports.$ref_(u),
    array: exports.array_(u),
    tuple: exports.tuple_(u),
    allOf: exports.allOf_(u),
    anyOf: exports.anyOf_(u),
    oneOf: exports.oneOf_(u),
    not: exports.not_(u),
    dictionary: exports.dictionary_(u),
    type: exports.type_(u),
    object: exports.object_(u),
    top: exports.top,
    extend: exports.extend
}); };
