// unmock-fetch/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fetch_1 = require("./fetch");
exports.buildFetch = fetch_1.default;
let fetchInterceptor;
class FetchInterceptor {
    constructor(onSerializedRequest) {
        this.fetch = fetch_1.default(onSerializedRequest);
        if (typeof global !== "undefined") {
            this.originalFetch = {
                where: global,
                fetch: global.fetch,
            };
            global.fetch = this.fetch;
        }
        else if (typeof window !== "undefined") {
            this.originalFetch = { where: window, fetch: window.fetch };
            window.fetch = this.fetch;
        }
    }
    disable() {
        if (this.originalFetch) {
            this.originalFetch.where.fetch = this.originalFetch.fetch;
            this.originalFetch = undefined;
        }
    }
}
exports.FetchInterceptor = FetchInterceptor;
exports.default = {
    on(onSerializedRequest) {
        this.off();
        fetchInterceptor = new FetchInterceptor(onSerializedRequest);
        return fetchInterceptor;
    },
    off() {
        if (fetchInterceptor) {
            fetchInterceptor.disable();
            fetchInterceptor = undefined;
        }
    },
};
