// io-ts/lib/Encoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.categoryEncoder = exports.contravariantEncoder = exports.URI = exports.compose = exports.contramap = exports.lazy = exports.sum = exports.intersection = exports.tuple = exports.array = exports.record = exports.partial = exports.type = exports.nullable = exports.id = void 0;
var Decoder_1 = require("./Decoder");
var Schemable_1 = require("./Schemable");
var function_1 = require("fp-ts/lib/function");
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
function id() {
    return {
        encode: function_1.identity
    };
}
exports.id = id;
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
function nullable(or) {
    return {
        encode: function (a) { return (a === null ? null : or.encode(a)); }
    };
}
exports.nullable = nullable;
/**
 * @since 2.2.3
 */
function type(properties) {
    return {
        encode: function (a) {
            var o = {};
            for (var k in properties) {
                o[k] = properties[k].encode(a[k]);
            }
            return o;
        }
    };
}
exports.type = type;
/**
 * @since 2.2.3
 */
function partial(properties) {
    return {
        encode: function (a) {
            var o = {};
            for (var k in properties) {
                var v = a[k];
                // don't add missing properties
                if (k in a) {
                    // don't strip undefined properties
                    o[k] = v === undefined ? undefined : properties[k].encode(v);
                }
            }
            return o;
        }
    };
}
exports.partial = partial;
/**
 * @since 2.2.3
 */
function record(codomain) {
    return {
        encode: function (r) {
            var o = {};
            for (var k in r) {
                o[k] = codomain.encode(r[k]);
            }
            return o;
        }
    };
}
exports.record = record;
/**
 * @since 2.2.3
 */
function array(items) {
    return {
        encode: function (as) { return as.map(items.encode); }
    };
}
exports.array = array;
/**
 * @since 2.2.3
 */
function tuple() {
    var components = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        components[_i] = arguments[_i];
    }
    return {
        encode: function (as) { return components.map(function (c, i) { return c.encode(as[i]); }); }
    };
}
exports.tuple = tuple;
/**
 * @since 2.2.3
 */
function intersection(left, right) {
    return {
        encode: function (ab) { return Decoder_1.intersect(left.encode(ab), right.encode(ab)); }
    };
}
exports.intersection = intersection;
/**
 * @since 2.2.3
 */
function sum(tag) {
    return function (members) {
        return {
            encode: function (a) { return members[a[tag]].encode(a); }
        };
    };
}
exports.sum = sum;
/**
 * @since 2.2.3
 */
function lazy(f) {
    var get = Schemable_1.memoize(f);
    return {
        encode: function (a) { return get().encode(a); }
    };
}
exports.lazy = lazy;
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.contramap = function (f) { return function (fa) {
    return contramap_(fa, f);
}; };
var contramap_ = function (fa, f) { return ({
    encode: function (b) { return fa.encode(f(b)); }
}); };
/**
 * @since 2.2.3
 */
exports.compose = function (ea) { return function (ab) {
    return compose_(ab, ea);
}; };
var compose_ = function (ab, ea) { return ({
    encode: function (b) { return ea.encode(ab.encode(b)); }
}); };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.URI = 'io-ts/Encoder';
/**
 * @since 2.2.3
 */
exports.contravariantEncoder = {
    URI: exports.URI,
    contramap: contramap_
};
/**
 * @since 2.2.3
 */
exports.categoryEncoder = {
    URI: exports.URI,
    compose: compose_,
    id: id
};
