// io-ts/lib/JsonEncoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemableJsonEncoder = exports.contravariantJsonEncoder = exports.URI = exports.contramap = exports.lazy = exports.sum = exports.intersection = exports.tuple = exports.array = exports.record = exports.partial = exports.type = exports.nullable = exports.id = void 0;
var function_1 = require("fp-ts/lib/function");
var E = require("./Encoder");
// -------------------------------------------------------------------------------------
// primitives
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.id = {
    encode: function_1.identity
};
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.nullable = E.nullable;
/**
 * @since 2.2.3
 */
exports.type = E.type;
/**
 * @since 2.2.3
 */
exports.partial = E.partial;
/**
 * @since 2.2.3
 */
exports.record = E.record;
/**
 * @since 2.2.3
 */
exports.array = E.array;
/**
 * @since 2.2.3
 */
exports.tuple = E.tuple;
/**
 * @since 2.2.3
 */
exports.intersection = E.intersection;
/**
 * @since 2.2.3
 */
exports.sum = E.sum;
/**
 * @since 2.2.3
 */
exports.lazy = E.lazy;
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.contramap = E.contramap;
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.URI = 'io-ts/JsonEncoder';
/**
 * @since 2.2.3
 */
exports.contravariantJsonEncoder = {
    URI: exports.URI,
    contramap: E.contravariantEncoder.contramap
};
/**
 * @since 2.2.3
 */
exports.schemableJsonEncoder = {
    URI: exports.URI,
    literal: function () { return exports.id; },
    string: exports.id,
    number: exports.id,
    boolean: exports.id,
    nullable: exports.nullable,
    type: exports.type,
    partial: exports.partial,
    record: exports.record,
    array: exports.array,
    tuple: exports.tuple,
    intersection: exports.intersection,
    sum: exports.sum,
    lazy: function (_, f) { return exports.lazy(f); }
};
