// io-ts/lib/JsonCodec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemableJsonCodec = exports.invariantJsonCodec = exports.URI = exports.imap = exports.lazy = exports.sum = exports.intersection = exports.tuple = exports.array = exports.record = exports.partial = exports.type = exports.nullable = exports.refinement = exports.withExpected = exports.boolean = exports.number = exports.string = exports.literal = exports.make = void 0;
var C = require("./Codec");
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.make = C.make;
/**
 * @since 2.2.3
 */
exports.literal = C.literal;
// -------------------------------------------------------------------------------------
// primitives
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.string = C.string;
/**
 * @since 2.2.3
 */
exports.number = C.number;
/**
 * @since 2.2.3
 */
exports.boolean = C.boolean;
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.withExpected = C.withExpected;
/**
 * @since 2.2.3
 */
exports.refinement = C.refinement;
/**
 * @since 2.2.3
 */
exports.nullable = C.nullable;
/**
 * @since 2.2.3
 */
exports.type = C.type;
/**
 * @since 2.2.3
 */
exports.partial = C.partial;
/**
 * @since 2.2.3
 */
exports.record = C.record;
/**
 * @since 2.2.3
 */
exports.array = C.array;
/**
 * @since 2.2.3
 */
exports.tuple = C.tuple;
/**
 * @since 2.2.3
 */
exports.intersection = C.intersection;
/**
 * @since 2.2.3
 */
exports.sum = C.sum;
/**
 * @since 2.2.3
 */
exports.lazy = C.lazy;
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.imap = C.imap;
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.2.3
 */
exports.URI = 'io-ts/JsonCodec';
/**
 * @since 2.2.3
 */
exports.invariantJsonCodec = {
    URI: exports.URI,
    imap: C.invariantCodec.imap
};
/**
 * @since 2.2.3
 */
exports.schemableJsonCodec = {
    URI: exports.URI,
    literal: exports.literal,
    string: exports.string,
    number: exports.number,
    boolean: exports.boolean,
    nullable: exports.nullable,
    type: exports.type,
    partial: exports.partial,
    record: exports.record,
    array: exports.array,
    tuple: exports.tuple,
    intersection: exports.intersection,
    sum: exports.sum,
    lazy: exports.lazy,
    refinement: exports.refinement
};
